### ISSUE TITLE
False warnings on `Enum.with_index/2`

### ISSUE THREAD

**Issue by losttime** [rrrene/credo#416]
### Environment

* Credo version (`mix credo -v`): `0.8.4`
* Erlang/Elixir version (`elixir -v`): `1.5.1`
* Operating system: `macOS 10.12.6`

### What were you trying to do?

Use `Enum.with_index/2` as the first parameter of `Enum.each/2`, followed by basically anything

```
  defp do_something(things) do
    Enum.each(Enum.with_index(things, 1), fn({thing, index}) ->
      IO.puts("Thing #{index}")
    end)
    IO.puts("done")
  end
```

### Expected outcome

No credo Warnings

### Actual outcome

Credo Warning: `There should be no unused return values for Enum.with_index().` for that line

Without the `IO.puts/1` at the end, credo doesn't flag `Enum.with_index/2`, but with `IO.puts/1` (or anything else, it seems), credo flags `Enum.with_index/2`.

#### Examples:

Warns (followed by function call instead of `IO.puts/2`):
```
  defp do_something(things) do
    Enum.each(Enum.with_index(things, 1), fn({thing, index}) ->
      IO.puts("Thing #{index}")
    end)
    do_something_else(things)
  end
```

Doesn't Warn (in a case block):
```
  defp do_something(things) do
    case things do
      [] ->
        IO.puts("no things")
      validated_things ->
        Enum.each(Enum.with_index(things, 1), fn({thing, index}) ->
          IO.puts("Thing #{index}")
        end)
        do_something_else(things)
    end
  end
```

Still Warns (case block followed by function call) :
```
  defp do_something(things) do
    case things do
      [] ->
        IO.puts("no things")
      validated_things ->
        Enum.each(Enum.with_index(things, 1), fn({thing, index}) ->
          IO.puts("Thing #{index}")
        end)
        do_something_else(things)
    end
    do_another_thing(things)
  end
```

**Comment by rrrene** [comment]
Thx for reporting this! :+1:

**Comment by philipdexter** [comment]
This looks like the same issue as in #339.

This issue can be reduced to

```elixir
defmodule A do
  def a do
    Enum.each(Enum.with_index([], 1), fn a -> a end)
    1
  end
end
```

The following does not produce the warning

```elixir
defmodule A do
  def blah(_) do
    1
  end

  def a do
    Enum.each(Enum.with_index([], 1), &blah/1)
    1
  end
end
```
